src/sync.h

```diff
@@ -70,13 +70,34 @@ inline void EnterCritical(const char* pszName, const char* pszFile, int nLine, v
 inline void LeaveCritical() {}
 inline void CheckLastCritical(void* cs, std::string& lockname, const char* guardname, const char* file, int line) {}
 template <typename MutexType>
-inline void AssertLockHeldInternal(const char* pszName, const char* pszFile, int nLine, MutexType* cs) EXCLUSIVE_LOCKS_REQUIRED(cs) {}
+inline void AssertLockHeldInternal(const char* pszName, const char* pszFile, int nLine, MutexType* cs) {}
 template <typename MutexType>
 void AssertLockNotHeldInternal(const char* pszName, const char* pszFile, int nLine, MutexType* cs) EXCLUSIVE_LOCKS_REQUIRED(!cs) {}
 inline void DeleteLock(void* cs) {}
 inline bool LockStackEmpty() { return true; }
 #endif
-#define AssertLockHeld(cs) AssertLockHeldInternal(#cs, __FILE__, __LINE__, &cs)
+
+/**
+ * Assert at compile time and run time that a mutex is locked. Has no effect
+ * when EXCLUSIVE_LOCKS_REQUIRED annotations are enforced because the runtime
+ * assertions will never trigger (analogous to asserting an unsigned int is
+ * greator or equal to 0).
+ *
+ * Use of this macro is neither encouraged nor discouraged in new code. It was
+ * historically used before compile time checks were available, but it may
+ * still be be useful for virtual methods or std::function callbacks where
+ * EXCLUSIVE_LOCKS_REQUIRED annotations are not enforced.
+ */
+#define AssertLockHeld(mutex) [&]() EXCLUSIVE_LOCKS_REQUIRED(mutex) { AssertLockHeldInternal(#mutex, __FILE__, __LINE__, &mutex); }()
+
+/**
+ * Assert at runtime that a mutex is locked. May be necessary to work around
+ * EXCLUSIVE_LOCKS_REQUIRED errors when the compiler can't determine that a
+ * mutex is locked.
+ */
+#define WeaklyAssertLockHeld(mutex) [&]() ASSERT_EXCLUSIVE_LOCK(mutex) { AssertLockHeldInternal(#mutex, __FILE__, __LINE__, &mutex); }()
```

Comment:
Maybe `LOCK_ALREADY_HELD(mutex)` would be better, comparable with `LOCK(mutex)`? This isn't a particularly weak assertion in anyway I can see, it's just one the compiler isn't able to verify at compile time.

"Assert that a mutex was already locked in any possible way this code could be reached. If DEBUG_LOCKORDER is defined, a runtime check that the mutex was actually locked by this thread will be made." ? Without DEBUG_LOCKORDER this does nothing at runtime.

Note that ASSERT_EXCLUSIVE_LOCK has slightly weird scoping behaviour; if `x` requires mutex `m` then:

```c++
    if (1) { WeaklyAssertLockHeld(m); }
    ++x;
```

works fine, while

```c++
    extern int n;
    if (n > 1) { WeaklyAssertLockHeld(m); }
    ++x;
```
tells you you need to hold the mutex to do `++x`. I couldn't come up with a case where this results in missing a warning though; but I'm not super-confident in it, as compared to being tightly bounded by scopes the way `SCOPED_LOCKABLE` is.

Reply:
In non-debug build a code like:

```cpp
WeaklyAssertLockHeld(mutex_required_for_x);
x = 1;
```

where the mutex is not held will **not** produce a compilation warning and will **not** assert at runtime. So the unprotected access will actually happen. This is because `ASSERT_EXCLUSIVE_LOCK` tells the compiler we are going to check, but we actually don't check (in non-debug build). Maybe remove `ASSERT_EXCLUSIVE_LOCK` in non-debug builds, or at least clearly explain in the comment the dangers of using this.

Reply:
> This seems like a brain dump and I can't figure out the suggestion. Is it to rename one macro or rename two macros?

Just to rename "WeaklyAssertLockHeld" to "LOCK_ALREADY_HELD" or similar.

I don't think it's accurate to say one is "less-preferred" to the other -- they do different things, and are appropriate in different circumstances. Hopefully the circumstances where WeaklyAssertLockHeld is appropriate are more rare, is all.

> "Assert that a mutex was already locked in any possible way this code could be reached. If DEBUG_LOCKORDER is defined, a runtime check that the mutex was actually locked by this thread will be made."

was a suggestion regarding the code docs, since it only has any runtime effect when DEBUG_LOCKORDER is defined.

Reply:
> Note that ASSERT_EXCLUSIVE_LOCK has slightly weird scoping behaviour [...]

This is unrelated to scopes, the capability is assumed to be held after the assertion, regardless of whether we leave any scopes. (Think about it, there is no destructor for the assertion, so why would the scope ending change anything?)

This analysis works on the source-level CFG. We can inspect this by passing `--analyze -Xanalyzer -analyzer-checker=debug.DumpCFG` to Clang. For the first example (using the [vocabulary from the documentation](https://clang.llvm.org/docs/ThreadSafetyAnalysis.html#mutex-h)):

```
 [B4 (ENTRY)]
   Succs (1): B3

 [B1]
   1: x
   2: ++[B1.1]
   Preds (2): B2 B3(Unreachable)
   Succs (1): B0

 [B2]
   1: m
   2: [B2.1].AssertHeld
   3: [B2.2]()
   Preds (1): B3
   Succs (1): B1

 [B3]
   1: 1
   2: [B3.1] (ImplicitCastExpr, IntegralToBoolean, _Bool)
   T: if [B3.2]
   Preds (1): B4
   Succs (2): B2 B1(Unreachable)

 [B0 (EXIT)]
   Preds (1): B1
```

The `if` is in `B3` and branches on a compile-time constant, thus we see that the branch to `B1` (where the increment happens) is unreachable. In your second example the branch condition is no longer a compile-time constant. (Try adding `const` or `constexpr`, then it should work.)

> This is because `ASSERT_EXCLUSIVE_LOCK` tells the compiler we are going to check, but we actually don't check (in non-debug build).

Using `ASSERT_EXCLUSIVE_LOCK` even though you don't always check is fine. It would be strange if it wasn't, because that's how the standard [assert](https://en.cppreference.com/w/cpp/error/assert) behaves.

> Maybe remove `ASSERT_EXCLUSIVE_LOCK` in non-debug builds, or at least clearly explain in the comment the dangers of using this.

Having static analysis depend on the build profile doesn't really make sense IMO. If you don't have the same annotations in all profiles you're just going to make your life harder.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/19918#discussion_r488282255

> Just to rename "WeaklyAssertLockHeld" to "LOCK_ALREADY_HELD" or similar.

I'm still not understanding why this is a good idea, but I made a space for it https://github.com/bitcoin-core/bitcoin-devwiki/wiki/AssertLockHeld-PRs